### PR TITLE
[BEAM-986] Deprecate TimerCallback and InMemoryTimerInternals methods using it.

### PR DESCRIPTION
**Pull request by scwhittle**
Deprecate TimerCallback and InMemoryTimerInternals methods using it.
Instead separate advancing watermarks and removing eligible timers.
SDK changes necessary for improving ReduceFnRunner prefetching.

Hi @kennknowles , can you please take a look?  This is the sdk changes separated out from PR adding
prefetching.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/InMemoryTimerInternals.java [hunk 1]
```diff
@@ -144,27 +150,26 @@ public String toString() {
     return MoreObjects.toStringHelper(getClass())
         .add("watermarkTimers", watermarkTimers)
         .add("processingTimers", processingTimers)
+        .add("synchronizedProcessingTimers", processingTimers)
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/InMemoryTimerInternals.java]: should be your new `synchronizedProcessingTimers`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/InMemoryTimerInternals.java [hunk 2]
```diff
@@ -144,27 +150,26 @@ public String toString() {
...
+  public void advanceInputWatermark(Instant newInputWatermark) throws Exception {
     checkNotNull(newInputWatermark);
     checkState(
         !newInputWatermark.isBefore(inputWatermarkTime),
         "Cannot move input watermark time backwards from %s to %s",
         inputWatermarkTime,
         newInputWatermark);
     WindowTracing.trace(
-        "TestTimerInternals.advanceInputWatermark: from {} to {}",
+        "InMemoryTimerInternals.advanceInputWatermark: from {} to {}",
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/InMemoryTimerInternals.java]: Awkward that this message made it in to the codebase. Can we use `getClass().getSimpleName()`?

**Comment by scwhittle** [comment]
@kennknowles Thanks! addressed you comments

**Comment by kennknowles** [comment]
Merged. Thanks for the very nice code!

I interpreted your commit messages and their contents as actually being two useful commits (independent rollback, independent history, etc) thusly: the uncapitalized commits I squashed and the capitalized commits I kept.

In the future, should you desire, you could explicitly signal what I should squash where using `fixup!` commit messages, which are automated via `git commit --fixup <hash>`. I am also not that sensitive to mid-review squashes; GitHub's UI has improved with regards to comments disappearing.

**Pull request closed without merging**